# BELAJAR PYTHON TEKOM F 2024

## Praktikum Dikelas

In [ ]:
print("Nama : Syahruramadan Kahar")
print("NIM : 240210501012")

Nama : Syahruramadan Kahar
NIM : 240210501012


### Analisis:
Kode tersebut menggunakan dua perintah print() untuk menampilkan nama dan NIM ke layar.

## Percobaan 1 Comments

In [5]:
#Creating a Comment
#This is a comment
print("Hello, World!")

print("Hello, World!") #This is a comment

#print("Hello, World!")
print("Cheers, Mate!")



#Multiline Comments
#This is a comment
#written in
#more than just one line
print("Hello, World!")

"""
This is a comment
written in
more than just one line
"""
print("Hello, World!")

Hello, World!
Hello, World!
Cheers, Mate!
Hello, World!
Hello, World!


### Analisis
Kode program tersebut menunjukkan penggunaan komentar dalam Python, baik komentar satu baris menggunakan tanda pagar (#) maupun komentar multilines menggunakan tiga tanda kutip (""" ... """). Komentar digunakan untuk memberikan penjelasan pada kode tanpa memengaruhi jalannya program. Beberapa baris print() digunakan untuk menampilkan teks ke layar, sementara baris lain yang diawali # tidak dijalankan karena dianggap sebagai komentar.

## Percobaan 2 Variables

### 1.Python Variabel

In [83]:
#Creating Variables
x = 5
y = "John"
print(x)
print(y)

x = 4       # x is of type int
x = "Sally" # x is now of type str
print(x)

5
John
Sally


### Analisis:
Kode tersebut menunjukkan cara mendefinisikan variabel dalam Python dan perubahan tipe datanya. Variabel x awalnya bernilai 5 sebagai tipe integer, lalu y berisi teks “John” sebagai tipe string, dan keduanya ditampilkan menggunakan print(). Selanjutnya, x diubah menjadi string “Sally”, menunjukkan bahwa Python bersifat dynamically typed, sehingga tipe data variabel dapat berubah sesuai nilai yang diberikan.

In [82]:
#Casting
x = str(3)    # x will be '3'
y = int(3)    # y will be 3
z = float(3)  # z will be 3.0

### Analisis:
Kode tersebut menunjukkan proses konversi tipe data pada Python. Angka 3 diubah menjadi string menggunakan str(3) sehingga x bernilai '3', kemudian int(3) menghasilkan nilai integer 3 untuk y, dan float(3) menghasilkan nilai pecahan 3.0 untuk z.

In [ ]:
#Get the Type
x = 5
y = "John"
print(type(x))
print(type(y))

<class 'int'>
<class 'str'>


### Analisis:
Kode tersebut digunakan untuk menampilkan tipe data dari variabel menggunakan fungsi type(). Variabel x diberi nilai 5 yang otomatis dikenali sebagai tipe data integer, sedangkan y berisi teks "John" yang merupakan tipe data string.

In [7]:
#Single or Double Quotes?
x = "John"
# is the same as
x = 'John'

### Analisis:
Kode tersebut menunjukkan bahwa dalam Python, teks (string) dapat ditulis menggunakan tanda kutip dua (" ") maupun tanda kutip satu (' '). Kedua cara tersebut memiliki fungsi yang sama dan menghasilkan tipe data string.

In [9]:
#Case-Sensitive
a = 4
A = "Sally"
#A will not overwrite a

### Analisis:
Kode tersebut menunjukkan bahwa Python bersifat case-sensitive, artinya huruf besar dan huruf kecil dianggap sebagai nama variabel yang berbeda. Variabel a diberikan nilai 4, sedangkan A diberikan nilai "Sally", dan keduanya disimpan secara terpisah. Dengan demikian, nilai pada A tidak akan menimpa atau menggantikan nilai pada a karena kedua variabel tersebut dianggap berbeda oleh Python.

### 2.Variabel Names

In [10]:
myvar = "John"
my_var = "John"
_my_var = "John"
myVar = "John"
MYVAR = "John"
myvar2 = "John"

### Analisis:
Kode tersebut menampilkan beberapa contoh penulisan nama variabel yang valid dalam Python. Semua variabel tersebut mengikuti aturan penamaan yang benar, seperti diawali huruf atau garis bawah, tidak mengandung spasi, serta dapat menggunakan huruf besar, huruf kecil, dan angka (selama angka tidak di awal). Perbedaan gaya penulisan seperti myvar, my_var, myVar, dan MYVAR menunjukkan variasi naming convention yang dapat digunakan, sementara Python tetap membedakan huruf besar dan kecil dalam penamaan variabel.

In [11]:
2myvar = "John"
my-var = "John"
my var = "John"

SyntaxError: invalid decimal literal (2185399839.py, line 1)

### Analisis:
Kode tersebut berisi contoh penulisan variabel yang tidak valid dalam Python. Ketiga contoh menyalahi aturan penamaan variabel, karena variabel tidak boleh dimulai dengan angka (2myvar), tidak boleh mengandung karakter khusus seperti tanda minus (my-var), dan tidak boleh mengandung spasi (my var). Jika dijalankan, kode tersebut akan menghasilkan error karena nama variabel tidak sesuai dengan aturan sintaks Python.

In [12]:
#Camel Case
myVariableName = "John"

#Pascal Case
MyVariableName = "John"

#Snake Case
my_variable_name = "John"

### Analisis:
Kode tersebut menunjukkan tiga gaya penulisan nama variabel yang benar dalam Python. Bentuk myVariableName menggunakan gaya camelCase, MyVariableName menggunakan PascalCase, dan my_variable_name menggunakan snake_case. Ketiganya valid dan dapat digunakan sesuai preferensi atau standar penulisan yang diikuti dalam sebuah proyek, dengan snake_case paling umum digunakan dalam Python.

### 3.Assign Multiple

In [13]:
#Many Values to Multiple Variables
x, y, z = "Orange", "Banana", "Cherry"
print(x)
print(y)
print(z)

Orange
Banana
Cherry


### Analisis:
Kode tersebut menunjukkan teknik multiple assignment pada Python, yaitu memberikan beberapa nilai ke beberapa variabel dalam satu baris. Variabel x, y, dan z masing-masing diberi nilai "Orange", "Banana", dan "Cherry". Ketika dicetak menggunakan print(), setiap variabel menampilkan nilai yang telah diberikan. Teknik ini membuat penulisan kode lebih ringkas dan efisien.

In [14]:
#One Value to Multiple Variables
x = y = z = "Orange"
print(x)
print(y)
print(z)

Orange
Orange
Orange


### Analisis:
Kode tersebut menampilkan teknik multiple assignment dengan satu nilai yang sama. Variabel x, y, dan z semuanya diisi dengan nilai yang sama, yaitu "Orange", dalam satu baris kode. Saat dicetak menggunakan print(), ketiga variabel tersebut akan menampilkan nilai "Orange".

In [15]:
#Unpack a Collection
fruits = ["apple", "banana", "cherry"]
x, y, z = fruits
print(x)
print(y)
print(z)

apple
banana
cherry


### Analisis:
Kode tersebut menunjukkan teknik unpacking nilai dari sebuah list ke dalam beberapa variabel. List fruits berisi tiga elemen, yaitu "apple", "banana", dan "cherry". Nilai-nilai tersebut kemudian dibagi (di-unpack) ke variabel x, y, dan z secara berurutan. Saat dicetak, masing-masing variabel akan menampilkan satu elemen dari list sesuai urutannya.

### 4.Output Variabel

In [16]:
x = "Python is awesome"
print(x)

x = "Python"
y = "is"
z = "awesome"
print(x, y, z)

x = "Python "
y = "is "
z = "awesome"
print(x + y + z)

x = 5
y = 10
print(x + y)

x = 5
y = "John"
print(x, y)

Python is awesome
Python is awesome
Python is awesome
15
5 John


### Analisis:
Kode tersebut memperlihatkan beberapa cara menampilkan data menggunakan print() serta contoh penggunaan operator + dan pemisahan nilai dengan koma. Pada contoh pertama, sebuah string dicetak langsung. Pada bagian kedua, tiga variabel string dicetak sekaligus menggunakan print(x, y, z) yang secara otomatis menambahkan spasi antar nilai. Selanjutnya, print(x + y + z) menggunakan operator + untuk melakukan concatenation atau penggabungan string. Pada contoh berikutnya, x + y digunakan untuk menjumlahkan dua angka, menghasilkan nilai numerik 15. Terakhir, ketika x berupa angka dan y berupa string, penggunaan print(x, y) tetap berhasil karena dipisahkan dengan koma, namun jika menggunakan + dalam kasus ini akan menyebabkan error karena Python tidak dapat menggabungkan angka dan string secara langsung.

In [17]:
x = 5
y = "John"
print(x + y)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

### Analisis:
Kode tersebut akan menghasilkan error karena mencoba menggunakan operator + untuk menggabungkan nilai dengan tipe data berbeda. Variabel x merupakan tipe integer, sedangkan y merupakan string. Python tidak dapat menjumlahkan atau menggabungkan angka dan teks menggunakan +, sehingga program akan menampilkan pesan kesalahan tipe data (TypeError).

### 5.Global Variabel

In [84]:
#Global Variables
x = "awesome"
def myfunc():
  print("Python is " + x)
myfunc()

x = "awesome"
def myfunc():
  x = "fantastic"
  print("Python is " + x)
myfunc()
print("Python is " + x)

Python is awesome
Python is fantastic
Python is awesome


### Analisis:
Kode tersebut menunjukkan perbedaan penggunaan variabel global dan local dalam Python. Pada bagian pertama, variabel x didefinisikan di luar fungsi dan digunakan di dalam fungsi, sehingga fungsi menampilkan “Python is awesome” karena x bersifat global. Pada bagian kedua, di dalam fungsi didefinisikan variabel x baru dengan nilai “fantastic”, sehingga saat fungsi dijalankan, nilai yang ditampilkan adalah “Python is fantastic”. Namun, variabel global x tidak berubah, sehingga saat dicetak di luar fungsi, hasilnya tetap “Python is awesome”.

In [85]:
#The global Keyword
def myfunc():
  global x
  x = "fantastic"
myfunc()
print("Python is " + x)

x = "awesome"
def myfunc():
  global x
  x = "fantastic"
myfunc()
print("Python is " + x)

Python is fantastic
Python is fantastic


### Analisis:
Kode tersebut menunjukkan penggunaan kata kunci global untuk mengubah nilai variabel yang didefinisikan di luar fungsi. Pada contoh pertama, variabel x belum memiliki nilai sebelum fungsi dipanggil, namun karena menggunakan global x, fungsi menetapkan nilai "fantastic" sebagai variabel global. Hasil cetak menunjukkan “Python is fantastic”. Pada contoh kedua, meskipun x awalnya bernilai "awesome", fungsi kembali mengubah nilai global x menjadi "fantastic", sehingga hasil akhirnya tetap mencetak “Python is fantastic”.

## Percobaan 3 Type Data

In [20]:
#Getting the Data Type
x = 5
print(type(x))

<class 'int'>


### Analisis:
Kode tersebut menggunakan fungsi type() untuk mengetahui tipe data dari variabel. Variabel x diberi nilai 5, sehingga ketika dicetak dengan print(type(x)), program akan menampilkan bahwa x bertipe integer (int).

## Percobaan 4 Numbers

In [21]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'float'>
<class 'complex'>


### Analisis:
Kode tersebut menunjukkan tiga jenis tipe data numerik dalam Python, yaitu int, float, dan complex. Variabel x berisi bilangan bulat (int), y berisi bilangan desimal (float), dan z berisi bilangan kompleks (complex) yang ditandai dengan akhiran j. Melalui print(type(...)), program menampilkan tipe data masing-masing variabel, sehingga memperlihatkan bahwa Python dapat menangani berbagai jenis bilangan secara otomatis.

In [22]:
#Int
x = 1
y = 35656222554887711
z = -3255522

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'int'>
<class 'int'>


### Analisis:
Kode tersebut menunjukkan bahwa tipe data integer (int) pada Python dapat menampung angka dengan berbagai ukuran, baik kecil, sangat besar, maupun bernilai negatif. Variabel x berisi angka kecil, y berisi angka dengan jumlah digit yang sangat besar, dan z berisi angka negatif. Ketika setiap variabel diperiksa menggunakan type(), hasilnya tetap menunjukkan tipe int, menandakan bahwa Python tidak memiliki batasan khusus untuk ukuran bilangan integer selain kemampuan memori komputer.

In [23]:
#Float
x = 1.10
y = 1.0
z = -35.59

print(type(x))
print(type(y))
print(type(z))

x = 35e3
y = 12E4
z = -87.7e100

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>


### Analisis:
Kode tersebut menunjukkan penggunaan tipe data float dalam Python, baik dalam bentuk angka desimal biasa maupun dalam notasi ilmiah (scientific notation). Pada bagian pertama, variabel x, y, dan z diisi dengan angka desimal, termasuk nilai negatif, dan semuanya dikenali sebagai tipe float. Pada bagian kedua, nilai diberikan menggunakan notasi ilmiah seperti 35e3, 12E4, dan -87.7e100, yang juga menghasilkan tipe float, di mana huruf e atau E menandakan pangkat 10.

In [24]:
#Complex
x = 3+5j
y = 5j
z = -5j

print(type(x))
print(type(y))
print(type(z))

<class 'complex'>
<class 'complex'>
<class 'complex'>


### Analisis:
Kode tersebut menunjukkan penggunaan tipe data complex dalam Python. Variabel x, y, dan z berisi bilangan kompleks yang ditandai dengan akhiran j sebagai bagian imajiner. Nilai x memiliki bagian real dan imajiner, sedangkan y dan z hanya memiliki bagian imajiner positif dan negatif. Ketika dicek menggunakan type(), ketiganya dikenali sebagai tipe data complex, menunjukkan bahwa Python mendukung operasi dengan bilangan kompleks secara bawaan.

In [25]:
#Type Conversion
x = 1    # int
y = 2.8  # float
z = 1j   # complex

#convert from int to float:
a = float(x)

#convert from float to int:
b = int(y)

#convert from int to complex:
c = complex(x)

print(a)
print(b)
print(c)

print(type(a))
print(type(b))
print(type(c))

1.0
2
(1+0j)
<class 'float'>
<class 'int'>
<class 'complex'>


### Analisis:
Kode tersebut menunjukkan proses konversi tipe data numerik dalam Python. Variabel x, y, dan z masing-masing bertipe int, float, dan complex. Nilai x dikonversi menjadi float dan disimpan pada variabel a, nilai y dikonversi menjadi int dan disimpan pada b, serta x juga dikonversi menjadi complex dan disimpan pada c. Hasil konversi tersebut kemudian dicetak, beserta tipe datanya menggunakan type().

In [26]:
#Random Number
import random

print(random.randrange(1, 10))

8


### Analisis:
Kode tersebut menggunakan modul random untuk menghasilkan angka acak. Fungsi random.randrange(1, 10) akan menghasilkan sebuah angka acak dari rentang 1 hingga 9 (angka 10 tidak termasuk).

## Percobaan 5 Casting

In [27]:
#Specify a Variable Type

#Integers:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3

### Analisis:
Kode tersebut memperlihatkan penggunaan fungsi int() untuk mengonversi berbagai jenis nilai menjadi tipe data integer. Nilai 1 tetap menjadi integer, nilai 2.8 dibulatkan ke bawah menjadi 2, dan string "3" diubah menjadi angka 3.

In [28]:
#Floats:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

### Analisis:
Kode tersebut menunjukkan penggunaan fungsi float() untuk mengonversi nilai menjadi tipe data float. Nilai 1 diubah menjadi 1.0, nilai 2.8 tetap menjadi 2.8, sedangkan string "3" dan "4.2" juga berhasil dikonversi menjadi angka pecahan (3.0 dan 4.2).

In [30]:
#Strings:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

### Analisis:
Kode tersebut memperlihatkan penggunaan fungsi str() untuk mengonversi nilai menjadi tipe data string. Nilai "s1" tetap menjadi string, angka 2 diubah menjadi "2", dan angka desimal 3.0 diubah menjadi "3.0".

## Percobaan 6 Strings

### 1. Python String

In [4]:
#Strings
print("Hello")
print('Hello')

Hello
Hello


### Analisis:
Kode tersebut menunjukkan bahwa Python mengizinkan penulisan string menggunakan tanda kutip dua (" ") maupun tanda kutip satu (' '). Keduanya memiliki fungsi yang sama untuk menampilkan teks

In [1]:
#Quotes Inside Quotes
print("It's alright")
print("He is called 'Johnny'")
print('He is called "Johnny"')

It's alright
He is called 'Johnny'
He is called "Johnny"


### Analisis:
Kode tersebut menunjukkan penggunaan tanda kutip dalam penulisan string di Python. Dengan memanfaatkan perbedaan antara kutip satu (') dan kutip dua ("), teks yang mengandung tanda kutip dapat dicetak tanpa error.

In [2]:
#Assign String to a Variable
a = "Hello"
print(a)

Hello


### Analisis:
Kode tersebut mendefinisikan variabel a dengan nilai string "Hello", kemudian mencetak isi variabel tersebut menggunakan fungsi print().

In [3]:
#Multiline Strings
#double quotes
a = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print(a)

#single quotes
a = '''Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.'''
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.
Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


### Analisis:
Kode tersebut menunjukkan penggunaan tanda kutip tiga (""" atau ''') untuk membuat string multilines di Python. Dengan cara ini, teks dapat ditulis dalam beberapa baris tanpa perlu menambahkan karakter khusus.

In [4]:
#Strings are Arrays
a = "Hello, World!"
print(a[1])

e


### Analisis:
Kode tersebut menunjukkan cara mengakses karakter dalam string menggunakan indeks. Variabel a berisi string "Hello, World!", dan a[1] mengambil karakter pada indeks ke-1 (indeks dimulai dari 0). Karena indeks 1 adalah huruf 'e', maka program akan mencetak huruf e.

In [5]:
#Looping Through a String
for x in "banana":
  print(x)

b
a
n
a
n
a


### Analisis:
Kode tersebut menunjukkan cara melakukan **iterasi pada string** menggunakan perulangan `for`. String `"banana"` diperlakukan sebagai kumpulan karakter, sehingga perulangan akan menelusuri setiap huruf satu per satu. Pada setiap iterasi, variabel `x` menyimpan satu karakter dan kemudian dicetak. Hasilnya, setiap huruf pada kata “banana” akan tampil pada baris terpisah.

In [6]:
#Looping Through a String
a = "Hello, World!"
print(len(a))

13


### Analisis:
Kode tersebut menggunakan fungsi len() untuk menghitung jumlah karakter dalam sebuah string. Variabel a berisi teks "Hello, World!", dan len(a) akan menghasilkan nilai panjang string tersebut, termasuk spasi dan tanda baca.

In [7]:
#Check String
txt = "The best things in life are free!"
print("free" in txt)


txt = "The best things in life are free!"
if "free" in txt:
  print("Yes, 'free' is present.")

True
Yes, 'free' is present.


### Analisis:
Kode tersebut menunjukkan penggunaan operator in untuk memeriksa apakah suatu kata terdapat dalam sebuah string. Pada baris pertama, "free" in txt akan menghasilkan nilai True karena kata free ada dalam teks. Pada contoh kedua, pengecekan dilakukan dalam pernyataan if, sehingga jika kata free ditemukan di dalam string, program akan mencetak pesan "Yes, 'free' is present.".

In [8]:
#Check if NOT
txt = "The best things in life are free!"
print("expensive" not in txt)


txt = "The best things in life are free!"
if "expensive" not in txt:
  print("No, 'expensive' is NOT present.")

True
No, 'expensive' is NOT present.


### Analisis:
Kode tersebut menggunakan operator not in untuk memeriksa apakah sebuah kata tidak terdapat dalam string. Pada contoh pertama, "expensive" not in txt akan menghasilkan True karena kata expensive memang tidak ada dalam teks. Pada contoh kedua, pengecekan dilakukan dalam struktur if, sehingga ketika kata tersebut tidak ditemukan di dalam string, program menampilkan pesan "No, 'expensive' is NOT present.".

### 2. Slicing String

In [9]:
#Slicing
b = "Hello, World!"
print(b[2:5])

llo


### Analisis:
Kode tersebut menunjukkan cara melakukan slicing pada string untuk mengambil sebagian karakter berdasarkan indeks. Variabel b berisi string "Hello, World!", dan perintah b[2:5] akan mengambil karakter mulai dari indeks ke-2 hingga sebelum indeks ke-5. Dengan demikian, hasil yang dicetak adalah "llo", karena slicing mengambil karakter pada indeks 2, 3, dan 4.

In [10]:
#Slice From the Start
b = "Hello, World!"
print(b[:5])

Hello


### Analisis:
Kode tersebut menggunakan teknik slicing tanpa menentukan indeks awal. Pada string "Hello, World!", perintah b[:5] berarti mengambil karakter dari awal string hingga sebelum indeks ke-5. Hasilnya adalah "Hello", karena karakter pada indeks 0 sampai 4 akan diambil.

In [ ]:
#Slice To the End
b = "Hello, World!"
print(b[2:])

### Analisis:
Kode tersebut menggunakan teknik slicing untuk mengambil bagian string mulai dari indeks tertentu hingga akhir. Pada string "Hello, World!", perintah b[2:] berarti mengambil karakter mulai dari indeks ke-2 sampai karakter terakhir. Hasil yang dicetak adalah "llo, World!".

In [11]:
#Negative Indexing
b = "Hello, World!"
print(b[-5:-2])

orl


### Analisis:
Kode tersebut menunjukkan penggunaan slicing dengan indeks negatif pada string. Indeks negatif menghitung posisi karakter dari belakang string. Pada string "Hello, World!", perintah b[-5:-2] mengambil karakter mulai dari indeks -5 hingga sebelum indeks -2. Hasilnya adalah "orl", karena karakter yang diambil adalah 'o', 'r', dan 'l'.

### 3. Modify String

In [12]:
#Upper Case
a = "Hello, World!"
print(a.upper())

HELLO, WORLD!


### Analisis:
Kode tersebut menggunakan metode upper() untuk mengubah semua huruf dalam string menjadi huruf kapital. Variabel a berisi teks "Hello, World!", dan ketika a.upper() dicetak, hasilnya akan menjadi "HELLO, WORLD!".

In [13]:
#Lower Case
a = "Hello, World!"
print(a.lower())

hello, world!


### Analisis:
Kode tersebut menggunakan metode lower() untuk mengubah semua huruf dalam string menjadi huruf kecil. Variabel a berisi teks "Hello, World!", dan saat a.lower() dicetak, hasilnya menjadi "hello, world!".

In [14]:
#Remove Whitespace
a = " Hello, World! "
print(a.strip()) # returns "Hello, World!"

Hello, World!


### Analisis:
Kode tersebut menggunakan metode strip() untuk menghapus spasi di bagian awal dan akhir string. Variabel a memiliki spasi sebelum dan sesudah teks, dan setelah a.strip() dijalankan, hasilnya menjadi "Hello, World!" tanpa spasi tambahan.

In [15]:
#Replace String
a = "Hello, World!"
print(a.replace("H", "J"))

Jello, World!


### Analisis:
Kode tersebut menggunakan metode replace() untuk mengganti karakter dalam string. Variabel a berisi "Hello, World!", dan perintah a.replace("H", "J") mengganti huruf H dengan J, sehingga hasil yang dicetak adalah "Jello, World!".

In [16]:
#Split String
a = "Hello, World!"
print(a.split(",")) # returns ['Hello', ' World!']

['Hello', ' World!']


### Analisis:
Kode tersebut menggunakan metode `split()` untuk memisahkan string berdasarkan tanda pemisah tertentu. Variabel `a` berisi `"Hello, World!"`, dan perintah `a.split(",")` memecah string menjadi dua bagian di setiap kemunculan koma. Hasilnya adalah sebuah list `['Hello', ' World!']`, yang berisi dua elemen teks terpisah.


### 4. Concatenate String

In [17]:
#String Concatenation
a = "Hello"
b = "World"
c = a + b
print(c)


a = "Hello"
b = "World"
c = a + " " + b
print(c)

HelloWorld
Hello World


### Analisis:
Kode tersebut menunjukkan cara menggabungkan string dengan operator +. Pada contoh pertama, variabel a dan b digabungkan langsung tanpa spasi, menghasilkan "HelloWorld". Pada contoh kedua, ditambahkan spasi di antara kedua string, sehingga output menjadi "Hello World".

### 5. Format String

In [18]:
#kita tidak dapat menggabungkan string dan angka seperti ini:
age = 36
#This will produce an error:
txt = "My name is John, I am " + age
print(txt)

TypeError: can only concatenate str (not "int") to str

### Analisis:
Kode tersebut akan menghasilkan error karena mencoba menggabungkan string dengan angka menggunakan operator +. Variabel age bertipe integer, sehingga tidak dapat langsung digabungkan dengan teks. Python hanya mengizinkan penggabungan menggunakan + jika semua bagian yang digabungkan bertipe string.

In [19]:
#F-Strings
age = 36
txt = f"My name is John, I am {age}"
print(txt)

My name is John, I am 36


### Analisis:
Kode tersebut menggunakan f-string untuk menggabungkan teks dengan variabel dengan cara yang lebih mudah dan rapi. Variabel age yang bernilai 36 disisipkan ke dalam string melalui {age} tanpa perlu melakukan konversi tipe secara manual. Ketika dicetak, program akan menampilkan: "My name is John, I am 36".

In [20]:
#Placeholders and Modifiers
price = 59
txt = f"The price is {price} dollars"
print(txt)


price = 59
txt = f"The price is {price:.2f} dollars"
print(txt)


txt = f"The price is {20 * 59} dollars"
print(txt)

The price is 59 dollars
The price is 59.00 dollars
The price is 1180 dollars


### Analisis:
Kode tersebut menunjukkan penggunaan f-string untuk memformat dan menampilkan nilai dalam string dengan lebih fleksibel. Pada contoh pertama, nilai variabel price langsung dimasukkan ke dalam teks sehingga menghasilkan "The price is 59 dollars". Pada contoh kedua, digunakan format .2f untuk menampilkan angka dalam dua angka di belakang koma, sehingga output menjadi "The price is 59.00 dollars". Pada contoh ketiga, f-string juga dapat memuat ekspresi matematika, dimana 20 * 59 akan dihitung terlebih dahulu sebelum ditampilkan, sehingga hasilnya adalah "The price is 1180 dollars".

### 6. Escape String

In [ ]:
txt = "We are the so-called "Vikings" from the north."

txt = "We are the so-called \"Vikings\" from the north."

### Analisis:
Kode tersebut menunjukkan cara menuliskan tanda kutip ganda di dalam string. Baris pertama akan menghasilkan error karena tanda kutip di sekitar kata "Vikings" dianggap sebagai penutup string. Pada baris kedua, digunakan karakter escape \" sehingga tanda kutip ganda dapat muncul sebagai bagian dari teks tanpa menutup string.

## Percobaan 7 Boolean

In [21]:
#Boolean Values
print(10 > 9)
print(10 == 9)
print(10 < 9)


a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")


#Evaluate Values and Variables
#Evaluate a string and a number:
print(bool("Hello"))
print(bool(15))

#Evaluate two variables:
x = "Hello"
y = 15
print(bool(x))
print(bool(y))

True
False
False
b is not greater than a
True
True
True
True


### Analisis:
Kode tersebut menunjukkan penggunaan operator perbandingan dan nilai kebenaran (Boolean) dalam Python. Bagian pertama mencetak hasil perbandingan angka sehingga menghasilkan nilai True atau False. Selanjutnya, contoh if-else digunakan untuk mengevaluasi kondisi apakah b lebih besar dari a, dan mencetak pesan sesuai hasilnya. Bagian berikutnya menunjukkan bahwa fungsi bool() akan menghasilkan True untuk string yang tidak kosong dan angka selain nol. Variabel x dan y juga diuji menggunakan bool(), dan keduanya menghasilkan True karena nilai yang diberikan tidak kosong atau nol.

In [22]:
#Most Values are True
bool("abc")
bool(123)
bool(["apple", "cherry", "banana"])

True

### Analisis:
Kode tersebut menunjukkan bahwa fungsi bool() akan mengembalikan nilai True untuk hampir semua data yang memiliki nilai atau isi. String "abc", angka 123, dan list ["apple", "cherry", "banana"] semuanya dianggap bernilai True karena tidak kosong.

In [23]:
#Some Values are False
bool(False)
bool(None)
bool(0)
bool("")
bool(())
bool([])
bool({})


class myclass():
  def __len__(self):
    return 0
myobj = myclass()
print(bool(myobj))

False


### Analisis:
Kode tersebut menunjukkan berbagai contoh nilai yang dianggap False oleh fungsi bool() di Python. Nilai seperti False, None, angka 0, string kosong "", tuple kosong (), list kosong [], dan dictionary kosong {} semuanya akan menghasilkan False karena tidak memiliki nilai atau isi. Pada contoh kelas myclass, metode __len__() mengembalikan nilai 0, sehingga objek myobj juga dianggap tidak memiliki panjang dan menghasilkan False ketika diuji dengan bool().

In [24]:
#Functions can Return a Boolean
def myFunction() :
  return True
print(myFunction())


def myFunction() :
  return True
if myFunction():
  print("YES!")
else:
  print("NO!")


x = 200
print(isinstance(x, int))

True
YES!
True


### Analisis:
Kode tersebut menunjukkan penggunaan fungsi yang mengembalikan nilai Boolean serta penggunaan fungsi isinstance(). Fungsi myFunction() selalu mengembalikan True, sehingga saat dicetak hasilnya adalah True, dan dalam struktur if-else, kondisi terpenuhi sehingga program mencetak "YES!". Pada bagian terakhir, isinstance(x, int) digunakan untuk memeriksa apakah variabel x merupakan tipe data int. Karena x bernilai 200 dan bertipe integer, fungsi tersebut mengembalikan True.

## Percobaan 8 Operator

### 1. Python Operator

In [25]:
print(10 + 5)

15


### Analisis:
Kode tersebut menampilkan hasil penjumlahan dua bilangan. Ekspresi 10 + 5 akan dihitung terlebih dahulu, kemudian print() menampilkan hasilnya ke layar. Output yang dihasilkan adalah 15.

In [26]:
sum1 = 100 + 50      # 150 (100 + 50)
sum2 = sum1 + 250    # 400 (150 + 250)
sum3 = sum2 + sum2   # 800 (400 + 400)

### Analisis:
Kode tersebut menunjukkan proses penjumlahan bertahap menggunakan variabel. Nilai awal sum1 adalah hasil dari 100 + 50 yaitu 150. Nilai tersebut kemudian digunakan untuk menghitung sum2, yaitu 150 + 250 yang menghasilkan 400. Selanjutnya, sum3 dihitung dengan menjumlahkan sum2 dengan dirinya sendiri, yaitu 400 + 400 sehingga menghasilkan 800.

### 2. Operator Aritmatika

In [27]:
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)
print(x ** y)
print(x // y)

#Division in Python
x = 12
y = 5
print(x / y)


x = 12
y = 5
print(x // y)

19
11
60
3.75
3
50625
3
2.4
2


### Analisis:
Kode tersebut menunjukkan penggunaan berbagai operator aritmatika dalam Python untuk melakukan perhitungan matematis. Pada bagian pertama, operasi antara x = 15 dan y = 4 mencakup penjumlahan (+), pengurangan (-), perkalian (*), pembagian (/), modulus (%) untuk sisa bagi, pangkat (**), dan pembagian bulat (//) yang menghasilkan nilai tanpa desimal. Contohnya, 15 / 4 menghasilkan nilai pecahan, sedangkan 15 // 4 menghasilkan hasil pembagian bulat. Pada bagian selanjutnya, ketika x = 12 dan y = 5, 12 / 5 menghasilkan nilai desimal, sedangkan 12 // 5 menghasilkan hasil pembagian tanpa desimal.

### 3. Operator Penugasan

In [28]:
#The Walrus Operator
numbers = [1, 2, 3, 4, 5]
count = len(numbers)
if count > 3:
    print(f"List has {count} elements")
if (count := len(numbers)) > 3:
    print(f"List has {count} elements")

List has 5 elements
List has 5 elements


### Analisis:
Kode tersebut menunjukkan dua cara untuk mengecek jumlah elemen dalam list dan mencetak pesan jika jumlahnya lebih dari tiga. Pada contoh pertama, panjang list disimpan dalam variabel count menggunakan len(numbers) lalu dicek dalam kondisi if. Pada contoh kedua, digunakan walrus operator := yang memungkinkan melakukan assignment dan pengecekan dalam satu baris, sehingga kode menjadi lebih ringkas.

### 4. Operator Perbandingan

In [29]:
x = 5
y = 3

print(x == y)
print(x != y)
print(x > y)
print(x < y)
print(x >= y)
print(x <= y)

False
True
True
False
True
False


### Analisis:
Kode tersebut menunjukkan penggunaan operator perbandingan dalam Python untuk membandingkan nilai dua variabel. Dengan x = 5 dan y = 3, hasil perbandingan yang dicetak adalah: x == y menghasilkan False karena nilainya tidak sama, x != y menghasilkan True, x > y menghasilkan True, x < y menghasilkan False, x >= y menghasilkan True, dan x <= y menghasilkan False.

In [30]:
#Chaining Comparison Operators
x = 5
print(1 < x < 10)
print(1 < x and x < 10)

True
True


### Analisis:
Kode tersebut menunjukkan dua cara untuk melakukan pengecekan rentang nilai pada sebuah variabel. Dengan x = 5, pernyataan 1 < x < 10 memeriksa apakah nilai x berada di antara 1 dan 10, dan hasilnya True. Contoh kedua menggunakan operator logika and untuk melakukan pengecekan yang sama, yaitu 1 < x and x < 10, dan hasilnya juga **True.

### 5. Operator Logika

In [31]:
x = 5
print(x > 0 and x < 10)


x = 5
print(x < 5 or x > 10)


x = 5
print(not(x > 3 and x < 10))

True
False
False


### Analisis:
Kode tersebut menunjukkan penggunaan operator logika (and, or, dan not) untuk mengevaluasi kondisi pada variabel x.

Pada contoh pertama, x > 0 and x < 10 akan menghasilkan True karena kedua kondisi terpenuhi (x bernilai 5).

Pada contoh kedua, x < 5 or x > 10 menghasilkan False karena tidak ada satu pun kondisi yang benar (5 tidak kurang dari 5 dan tidak lebih dari 10).

Pada contoh ketiga, not(x > 3 and x < 10) awalnya kondisi di dalam kurung bernilai True, namun karena ada kata kunci not, hasil akhirnya dibalik menjadi False.

### 6. Operator Identitas

In [32]:
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x
print(x is z)
print(x is y)
print(x == y)


x = ["apple", "banana"]
y = ["apple", "banana"]
print(x is not y)

True
False
True
True


### Analisis:
Pada bagian pertama, x dan z menunjuk ke objek list yang sama sehingga x is z menghasilkan True. Sementara itu, x dan y memiliki isi yang sama, tetapi merupakan objek berbeda di memori, sehingga x is y menghasilkan False, tetapi x == y menghasilkan True karena nilai/isinya sama.

Pada bagian terakhir, x is not y menghasilkan True karena x dan y memang bukan objek yang sama di memori, meskipun isi list mereka identik.

In [33]:
#Difference Between is and ==
x = [1, 2, 3]
y = [1, 2, 3]

print(x == y)
print(x is y)

True
False


### Analisis:
Kode tersebut memperlihatkan perbedaan antara operator `==` dan `is` dalam Python. Meskipun list `x` dan `y` memiliki nilai atau isi yang sama sehingga `x == y` menghasilkan **True**, keduanya merupakan objek yang berbeda di memori, sehingga `x is y` menghasilkan **False**. Dengan demikian, `==` digunakan untuk membandingkan kesamaan nilai, sedangkan `is` membandingkan identitas objek di memori.


### 7. Operator Keanggotaan

In [34]:
fruits = ["apple", "banana", "cherry"]
print("banana" in fruits)


fruits = ["apple", "banana", "cherry"]
print("pineapple" not in fruits)

True
True


### Analisis:
Kode tersebut menunjukkan penggunaan operator keanggotaan in dan not in pada list. Pada contoh pertama, "banana" in fruits menghasilkan True karena elemen banana terdapat dalam list. Pada contoh kedua, "pineapple" not in fruits juga menghasilkan True karena pineapple tidak ada dalam list. Operator ini digunakan untuk memeriksa apakah suatu nilai termasuk atau tidak dalam sebuah koleksi data seperti list.

In [35]:
#Membership in Strings
text = "Hello World"

print("H" in text)
print("hello" in text)
print("z" not in text)

True
False
True


### Analisis:
Kode tersebut menggunakan operator in dan not in untuk memeriksa keberadaan karakter atau teks dalam sebuah string. Ekspresi "H" in text menghasilkan True karena huruf H terdapat dalam string "Hello World", sementara "hello" in text menghasilkan False karena pengecekan bersifat case-sensitive dan tidak sama dengan "Hello". Ekspresi "z" not in text menghasilkan True karena huruf z memang tidak terdapat di dalam string tersebut.

### 8. Operator Bitwise

In [36]:
print(6 & 3)

print(6 | 3)

print(6 ^ 3)

2
7
5


### Analisis:
Kode tersebut menunjukkan penggunaan operator bitwise pada Python untuk memanipulasi nilai biner dari angka.

6 & 3 melakukan operasi AND biner antara 6 (110) dan 3 (011), menghasilkan 010 atau 2.

6 | 3 melakukan operasi OR biner dan menghasilkan 111 atau 7.

6 ^ 3 melakukan operasi XOR biner dan menghasilkan 101 atau 5.

Operator bitwise membandingkan setiap bit dari dua bilangan dan menghasilkan nilai baru berdasarkan aturan operasi bit yang digunakan.

### 9. Prioritas Operator

In [37]:
print((6 + 3) - (6 + 3))

print(100 + 5 * 3)

0
115


### Analisis:
Kode tersebut menunjukkan cara Python menghitung ekspresi aritmatika sesuai urutan operasi. Pada baris pertama, ekspresi (6 + 3) - (6 + 3) menghasilkan nilai 0 karena kedua bagian bernilai 9 lalu dikurangkan. Pada baris kedua, 100 + 5 * 3 mengikuti aturan prioritas operasi (perkalian didahulukan), sehingga 5 * 3 = 15 lalu ditambahkan ke 100 menjadi 115.

In [38]:
#Left-to-Right Evaluation
print(5 + 4 - 7 + 3)

5


### Analisis:
Kode tersebut menjalankan operasi aritmatika secara berurutan dari kiri ke kanan karena hanya melibatkan penjumlahan dan pengurangan yang memiliki tingkat prioritas sama. Perhitungan 5 + 4 - 7 + 3 menghasilkan nilai 5, sehingga program akan mencetak angka 5.

## Percobaan 9 If Else

### 1. If

In [39]:
a = 33
b = 200
if b > a:
  print("b is greater than a")

b is greater than a


### Analisis:
Kode tersebut menggunakan pernyataan if untuk melakukan pengecekan kondisi sederhana. Variabel a bernilai 33 dan b bernilai 200, kemudian kondisi b > a diuji. Karena nilai b lebih besar dari a, kondisi bernilai True, sehingga program mengeksekusi perintah di dalam blok if dan mencetak pesan "b is greater than a".

In [40]:
#Checking if a number is positive:
number = 15
if number > 0:
  print("The number is positive")

The number is positive


### Analisis:
Kode tersebut menggunakan pernyataan if untuk mengecek apakah suatu angka bernilai positif. Variabel number berisi nilai 15, dan karena 15 lebih besar dari 0, kondisi number > 0 bernilai True, sehingga program mengeksekusi kode di dalam blok if dan mencetak pesan "The number is positive".

In [41]:
#Indentation
a = 33
b = 200
if b > a:
print("b is greater than a") # you will get an error

IndentationError: expected an indented block after 'if' statement on line 3 (1443227691.py, line 4)

### Analisis:
Kode tersebut akan menghasilkan error karena baris print("b is greater than a") tidak memiliki indentasi yang benar. Dalam Python, blok kode setelah pernyataan if harus ditulis dengan indentasi, dan karena tidak diberi spasi atau tab, program tidak dapat mengenali bagian tersebut sebagai isi dari kondisi if.

In [42]:
#Multiple Statements in If Block
age = 20
if age >= 18:
  print("You are an adult")
  print("You can vote")
  print("You have full legal rights")

You are an adult
You can vote
You have full legal rights


### Analisis:
Kode tersebut menggunakan pernyataan if untuk mengecek apakah seseorang sudah memenuhi batas usia dewasa. Variabel age bernilai 20, dan karena 20 ≥ 18, kondisi age >= 18 bernilai True sehingga seluruh perintah di dalam blok if dijalankan.

In [43]:
#Using Variables in Conditions
is_logged_in = True
if is_logged_in:
  print("Welcome back!")

Welcome back!


### Analisis:
Kode tersebut menggunakan pernyataan if untuk mengecek status login pengguna. Variabel is_logged_in bernilai True, sehingga kondisi if is_logged_in terpenuhi dan program menjalankan perintah di dalam blok if, yaitu mencetak pesan "Welcome back!".

### 2. Elif

In [44]:
#The Elif Keyword
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")

a and b are equal


### Analisis:
Kode tersebut menggunakan struktur if–elif untuk mengevaluasi beberapa kondisi. Dengan nilai a = 33 dan b = 33, kondisi pertama b > a bernilai False, sehingga program memeriksa kondisi berikutnya a == b, yang bernilai True. Karena itu, program mencetak pesan "a and b are equal".

In [45]:
#Multiple Elif Statements
score = 75

if score >= 90:
  print("Grade: A")
elif score >= 80:
  print("Grade: B")
elif score >= 70:
  print("Grade: C")
elif score >= 60:
  print("Grade: D")

Grade: C


### Analisis:
Kode tersebut menggunakan struktur if–elif untuk menentukan nilai grade berdasarkan skor. Variabel score bernilai 75, sehingga kondisi pertama dan kedua tidak terpenuhi, namun kondisi score >= 70 bernilai True, sehingga program mencetak "Grade: C".

In [46]:
#Categorizing age groups:
age = 25

if age < 13:
  print("You are a child")
elif age < 20:
  print("You are a teenager")
elif age < 65:
  print("You are an adult")
elif age >= 65:
  print("You are a senior")

You are an adult


### Analisis:
Kode tersebut menggunakan struktur if–elif untuk mengelompokkan usia ke dalam kategori tertentu. Variabel age bernilai 25, sehingga kondisi pertama dan kedua tidak terpenuhi, tetapi kondisi age < 65 bernilai True, sehingga program mencetak "You are an adult".

In [47]:
#Day of the week checker:
day = 3

if day == 1:
  print("Monday")
elif day == 2:
  print("Tuesday")
elif day == 3:
  print("Wednesday")
elif day == 4:
  print("Thursday")
elif day == 5:
  print("Friday")
elif day == 6:
  print("Saturday")
elif day == 7:
  print("Sunday")


Wednesday


### Analisis:
Kode tersebut menggunakan struktur if–elif untuk mencetak nama hari berdasarkan nilai numerik yang diberikan. Dengan day = 3, kondisi day == 3 bernilai True, sehingga program mencetak "Wednesday".

### 3. Else

In [48]:
#The Else Keyword
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")

a is greater than b


### Analisis:
Kode tersebut menggunakan struktur if–elif–else untuk membandingkan dua nilai. Dengan a = 200 dan b = 33, kondisi pertama b > a bernilai False, dan kondisi kedua a == b juga False, sehingga program mengeksekusi blok else dan mencetak "a is greater than b".

In [49]:
#Else Without Elif
a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis:
Kode tersebut menggunakan struktur if–else untuk membandingkan nilai dua variabel. Dengan a = 200 dan b = 33, kondisi b > a bernilai False, sehingga program menjalankan blok else dan mencetak pesan "b is not greater than a".

In [50]:
#How Else Works
#Checking even or odd numbers:
number = 7

if number % 2 == 0:
  print("The number is even")
else:
  print("The number is odd")

The number is odd


### Analisis:
Kode tersebut menggunakan struktur if–else untuk menentukan apakah sebuah angka tergolong genap atau ganjil. Variabel number bernilai 7, lalu dilakukan pengecekan dengan number % 2 == 0. Karena sisa hasil bagi 7 dengan 2 bukan 0, kondisi tersebut bernilai False, sehingga program menjalankan blok else dan mencetak "The number is odd".

In [51]:
#Complete If-Elif-Else Chain
#Temperature classifier:
temperature = 22

if temperature > 30:
  print("It's hot outside!")
elif temperature > 20:
  print("It's warm outside")
elif temperature > 10:
  print("It's cool outside")
else:
  print("It's cold outside!")

It's warm outside


### Analisis:
Kode tersebut menggunakan struktur if–elif–else untuk memberikan pesan sesuai dengan kondisi suhu. Variabel temperature bernilai 22, sehingga kondisi temperature > 30 tidak terpenuhi, tetapi temperature > 20 bernilai True, sehingga program mencetak "It's warm outside".

In [52]:
#Else as Fallback
#Validating user input:
username = "Emil"

if len(username) > 0:
  print(f"Welcome, {username}!")
else:
  print("Error: Username cannot be empty")

Welcome, Emil!


### Analisis:
Kode tersebut menggunakan struktur if–else untuk mengecek apakah variabel username berisi teks atau tidak. Karena username bernilai "Emil" dan panjangnya lebih dari 0, kondisi len(username) > 0 bernilai True, sehingga program menampilkan pesan "Welcome, Emil!".

### 4. Shorthand If

In [53]:
#Short Hand If
a = 5
b = 2
if a > b: print("a is greater than b")

a is greater than b


### Analisis:
Kode tersebut menggunakan bentuk if satu baris (short-hand if) untuk mengeksekusi perintah langsung setelah kondisi. Karena nilai a (5) lebih besar daripada b (2), kondisi a > b bernilai True dan program mencetak "a is greater than b".

In [54]:
#Short Hand If ... Else
a = 2
b = 330
print("A") if a > b else print("B")

B


### Analisis:
Kode tersebut menggunakan short-hand if–else dalam satu baris. Dengan nilai a = 2 dan b = 330, kondisi a > b bernilai False, sehingga bagian setelah else dijalankan dan program mencetak "B".

In [55]:
#Assign a Value With If ... Else
a = 10
b = 20
bigger = a if a > b else b
print("Bigger is", bigger)

#The syntax follows this pattern:
variable = value_if_true if condition else value_if_false

Bigger is 20


NameError: name 'condition' is not defined

### Analisis:
Kode tersebut menggunakan short-hand if–else (ternary operator) untuk menentukan nilai variabel berdasarkan kondisi. Dengan a = 10 dan b = 20, ekspresi a if a > b else b memilih nilai yang lebih besar, yaitu b, sehingga variabel bigger bernilai 20 dan program mencetak "Bigger is 20". Baris terakhir menunjukkan format umum penulisan ternary operator di Python:
variable = value_if_true if condition else value_if_false.

In [56]:
#Multiple Conditions on One Line
a = 330
b = 330
print("A") if a > b else print("=") if a == b else print("B")

=


### Analisis:
Kode tersebut menggunakan ternary operator bertingkat untuk mengevaluasi beberapa kondisi dalam satu baris. Dengan a = 330 dan b = 330, kondisi pertama a > b bernilai False, sehingga dilanjutkan ke kondisi berikutnya a == b yang bernilai True, sehingga program mencetak "=".

In [57]:
#Practical Examples
#Finding the maximum of two numbers:
x = 15
y = 20
max_value = x if x > y else y
print("Maximum value:", max_value)

#Setting a default value:
username = ""
display_name = username if username else "Guest"
print("Welcome,", display_name)

Maximum value: 20
Welcome, Guest


### Analisis:
Kode tersebut menunjukkan penggunaan ternary operator untuk memilih nilai berdasarkan kondisi. Pada contoh pertama, x dan y dibandingkan, dan karena y lebih besar, max_value menyimpan nilai 20 dan program mencetak "Maximum value: 20". Pada contoh kedua, variabel username bernilai string kosong, sehingga kondisi setelah if dianggap False dan display_name diisi dengan "Guest". Program kemudian mencetak "Welcome, Guest".

### 5. Logical Operator

In [58]:
#The and Operator
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


### Analisis:
Kode tersebut menggunakan operator logika and untuk mengecek dua kondisi sekaligus. Dengan nilai a = 200, b = 33, dan c = 500, kondisi a > b bernilai True dan c > a juga True, sehingga kedua kondisi terpenuhi. Karena kedua syarat benar, program mengeksekusi perintah di dalam blok if dan mencetak "Both conditions are True".

In [59]:
#The or Operator
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

At least one of the conditions is True


### Analisis:
Kode tersebut menggunakan operator logika or untuk mengecek apakah minimal satu dari dua kondisi bernilai benar. Dengan a = 200, b = 33, dan c = 500, kondisi a > b bernilai True, sementara a > c bernilai False. Karena operator or hanya membutuhkan satu kondisi yang benar, program tetap mengeksekusi blok if dan mencetak "At least one of the conditions is True".

In [60]:
#The not Operator
a = 33
b = 200
if not a > b:
  print("a is NOT greater than b")

a is NOT greater than b


### Analisis:
Kode tersebut menggunakan operator not untuk membalik hasil kondisi. Dengan a = 33 dan b = 200, kondisi a > b sebenarnya False, namun karena dibalik dengan not, hasilnya menjadi True, sehingga program mengeksekusi perintah dalam blok if dan mencetak "a is NOT greater than b".

In [61]:
#Combining Multiple Operators
age = 25
is_student = False
has_discount_code = True

if (age < 18 or age > 65) and not is_student or has_discount_code:
  print("Discount applies!")

Discount applies!


### Analisis:
Kode tersebut menggunakan kombinasi operator logika or, and, dan not untuk menentukan apakah diskon berlaku. Dengan age = 25, is_student = False, dan has_discount_code = True, ekspresi (age < 18 or age > 65) bernilai False, not is_student bernilai True, dan has_discount_code juga True. Karena terdapat kondisi yang bernilai True dan operator or digunakan di antaranya, hasil keseluruhan ekspresi adalah True, sehingga program mencetak "Discount applies!".

In [62]:
#Using Parentheses for Clarity
temperature = 25
is_raining = False
is_weekend = True

if (temperature > 20 and not is_raining) or is_weekend:
  print("Great day for outdoor activities!")

Great day for outdoor activities!


### Analisis:
Kode tersebut menggunakan kombinasi operator logika untuk mengevaluasi kondisi cuaca dan waktu sebelum menampilkan rekomendasi. Dengan temperature = 25, is_raining = False, dan is_weekend = True, ekspresi (temperature > 20 and not is_raining) bernilai True, dan is_weekend juga True, sehingga keseluruhan kondisi menghasilkan True. Akibatnya, program mengeksekusi perintah dalam blok if dan mencetak "Great day for outdoor activities!".

In [63]:
#User authentication check:
username = "Tobias"
password = "secret123"
is_verified = True

if username and password and is_verified:
  print("Login successful")
else:
  print("Login failed")

#Range checking with logical operators:
score = 85

if score >= 0 and score <= 100:
  print("Valid score")
else:
  print("Invalid score")

Login successful
Valid score


### Analisis:
Kode pertama melakukan pengecekan login dengan menggunakan operator logika and, di mana login dinyatakan berhasil jika username, password, dan is_verified semuanya bernilai True. Karena semua kondisi terpenuhi (username dan password tidak kosong serta is_verified bernilai True), program akan mencetak "Login successful". Kode kedua memvalidasi nilai score dengan memeriksa apakah berada dalam rentang 0 hingga 100. Karena nilai score adalah 85 dan berada dalam rentang yang benar, program mencetak "Valid score".

### 6. Nested If

In [64]:
x = 41

if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")

Above ten,
and also above 20!


### Analisis:
Kode tersebut menunjukkan penggunaan nested if atau percabangan bersarang. Variabel x bernilai 41, sehingga kondisi pertama x > 10 bernilai True dan program mencetak "Above ten,". Karena kondisi tersebut terpenuhi, pengecekan dilanjutkan ke kondisi di dalamnya, yaitu x > 20, yang juga bernilai True, sehingga program mencetak "and also above 20!".

In [65]:
#How Nested If Works
age = 25
has_license = True

if age >= 18:
  if has_license:
    print("You can drive")
  else:
    print("You need a license")
else:
  print("You are too young to drive")

You can drive


### Analisis:
Kode tersebut menggunakan percabangan bersarang (nested if) untuk menentukan apakah seseorang boleh mengemudi. Variabel age bernilai 25, sehingga kondisi age >= 18 bernilai True, dan program melanjutkan ke pengecekan berikutnya. Karena has_license juga bernilai True, program mencetak "You can drive".

In [66]:
#Multiple Levels of Nesting
score = 85
attendance = 90
submitted = True

if score >= 60:
  if attendance >= 80:
    if submitted:
      print("Pass with good standing")
    else:
      print("Pass but missing assignment")
  else:
    print("Pass but low attendance")
else:
  print("Fail")

Pass with good standing


### Analisis:
Kode tersebut menggunakan nested if bertingkat untuk menentukan status kelulusan berdasarkan tiga kriteria: nilai, kehadiran, dan pengumpulan tugas. Dengan score = 85, attendance = 90, dan submitted = True, semua kondisi utama terpenuhi, sehingga program mencetak "Pass with good standing". Struktur ini memeriksa setiap syarat secara berurutan, dan hanya jika suatu syarat terpenuhi barulah syarat berikutnya diuji, sehingga penilaian menjadi lebih spesifik dan terperinci.

In [67]:
#Nested If vs Logical Operators
#This nested if:
temperature = 25
is_sunny = True

if temperature > 20:
  if is_sunny:
    print("Perfect beach weather!")

#Could also be written with and:
temperature = 25
is_sunny = True

if temperature > 20 and is_sunny:
  print("Perfect beach weather!")

Perfect beach weather!
Perfect beach weather!


### Analisis:
Kedua potongan kode tersebut menghasilkan output yang sama, yaitu mencetak "Perfect beach weather!" ketika suhu di atas 20 dan cuacanya cerah. Contoh pertama menggunakan nested if di mana kondisi is_sunny hanya diperiksa jika temperature > 20 bernilai True, sedangkan contoh kedua menggunakan operator logika and untuk memeriksa kedua kondisi dalam satu baris.

In [68]:
#Login validation with nested checks:
username = "Emil"
password = "python123"
is_active = True

if username:
  if password:
    if is_active:
      print("Login successful")
    else:
      print("Account is not active")
  else:
    print("Password required")
else:
  print("Username required")

#Grade calculation with nested logic:
score = 92
extra_credit = 5

if score >= 90:
  if extra_credit > 0:
    print("A+ grade")
  else:
    print("A grade")
elif score >= 80:
  print("B grade")
else:
  print("C grade or below")

Login successful
A+ grade


### Analisis:
Kode pertama menggunakan nested if untuk memvalidasi login berdasarkan tiga kondisi: username, password, dan status akun (is_active). Karena semua kondisi bernilai True, program mencetak "Login successful". Kode kedua menilai nilai akhir berdasarkan skor dan tambahan nilai. Karena score bernilai 92 dan memenuhi kondisi score >= 90, program memeriksa apakah terdapat extra_credit, dan karena nilainya lebih dari 0, program mencetak "A+ grade".

### 7. Pass Statement

In [70]:
a = 33
b = 200

if b > a:
  pass

### Analisis:
Kode tersebut menggunakan pernyataan pass di dalam blok if. Variabel a bernilai 33 dan b bernilai 200, sehingga kondisi b > a sebenarnya True, namun karena blok if hanya berisi pass, program tidak melakukan tindakan apa pun. Kata kunci pass digunakan sebagai placeholder untuk menjaga struktur kode tetap valid ketika belum ada perintah yang ingin dijalankan di dalam blok tersebut.

In [72]:
#Placeholder for future implementation:
age = 16

if age < 18:
  pass # TODO: Add underage logic later
else:
  print("Access granted")

### Analisis:
Kode tersebut menggunakan pass sebagai placeholder dalam blok if ketika kondisi age < 18 terpenuhi. Karena age bernilai 16, kondisi tersebut True, tetapi tidak ada aksi yang dijalankan karena pass hanya berfungsi menjaga blok tetap kosong dan valid. Komentar # TODO: Add underage logic later menunjukkan bahwa logika untuk pengguna di bawah umur akan ditambahkan kemudian. Jika usia tidak kurang dari 18, barulah program mengeksekusi blok else dan mencetak "Access granted".

In [73]:
#pass vs Comments
score = 85

if score > 90:
  # This is excellent
# This will raise an IndentationError

_IncompleteInputError: incomplete input (2330558437.py, line 5)

### Analisis:
Kode tersebut akan menimbulkan IndentationError karena setelah pernyataan if score > 90: terdapat komentar yang tidak berada dalam blok yang ter-indent. Python mengharuskan adanya blok kode yang menjorok ke dalam setelah sebuah kondisi if, meskipun isinya hanya komentar atau satu baris kode.

In [74]:
#This works correctly with pass:
score = 85

if score > 90:
  pass # This is excellent
print("Score processed")

Score processed


### Analisis:
Kode tersebut menggunakan pass untuk mengisi blok if agar tetap valid tanpa menjalankan aksi apa pun. Karena score = 85 tidak memenuhi kondisi score > 90, blok if dilewati dan program langsung menjalankan perintah berikutnya, yaitu mencetak "Score processed". Penggunaan pass di sini berguna sebagai placeholder ketika logika di dalam if akan ditambahkan nanti.

In [75]:
#pass with Multiple Conditions
value = 50

if value < 0:
  print("Negative value")
elif value == 0:
  pass # Zero case - no action needed
else:
  print("Positive value")

Positive value


### Analisis:
Kode tersebut menggunakan pass dalam blok elif sebagai placeholder ketika nilai value sama dengan 0, menandakan bahwa tidak ada aksi yang perlu dilakukan untuk kondisi tersebut. Dengan value = 50, kondisi value < 0 bernilai False, lalu value == 0 juga False, sehingga program menjalankan blok else dan mencetak "Positive value".

In [76]:
#pass in Other Contexts
def calculate_discount(price):
  pass # TODO: Implement discount logic

# Function exists but doesn't do anything yet

### Analisis:
Kode tersebut mendefinisikan sebuah fungsi calculate_discount(price) namun belum berisi logika apa pun karena hanya menggunakan pass sebagai placeholder. Hal ini menandakan bahwa fungsi tersebut sudah dibuat agar dapat dipanggil, tetapi logika perhitungan diskonnya akan diimplementasikan nanti.

## Percobaan 10 Match

In [77]:
day = 4
match day:
  case 1:
    print("Monday")
  case 2:
    print("Tuesday")
  case 3:
    print("Wednesday")
  case 4:
    print("Thursday")
  case 5:
    print("Friday")
  case 6:
    print("Saturday")
  case 7:
    print("Sunday")

Thursday


### Analisis:
Kode tersebut menggunakan struktur match–case, fitur seperti switch dalam Python, untuk mencetak nama hari berdasarkan nilai variabel day. Dengan day = 4, program akan mencocokkan nilai tersebut dengan case yang sesuai dan mengeksekusi perintah pada case 4, sehingga mencetak "Thursday".

In [78]:
# Default Value
day = 4
match day:
  case 6:
    print("Today is Saturday")
  case 7:
    print("Today is Sunday")
  case _:
    print("Looking forward to the Weekend")

Looking forward to the Weekend


### Analisis:
Kode tersebut menggunakan struktur match–case dengan default case menggunakan simbol _. Variabel day bernilai 4, dan karena tidak cocok dengan case 6 atau case 7, program menjalankan case _ sebagai opsi default dan mencetak "Looking forward to the Weekend". Case _ berfungsi seperti default pada switch di bahasa lain, yaitu dijalankan jika tidak ada case yang sesuai.

In [79]:
# Combine Values
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5:
    print("Today is a weekday")
  case 6 | 7:
    print("I love weekends!")

Today is a weekday


### Analisis:
Kode tersebut menggunakan match–case dengan penggabungan beberapa nilai dalam satu case menggunakan operator |. Dengan day = 4, nilai tersebut cocok dengan case 1 | 2 | 3 | 4 | 5, sehingga program mencetak "Today is a weekday". Teknik ini memudahkan penulisan kondisi untuk banyak nilai tanpa harus membuat beberapa case terpisah, sehingga lebih ringkas dan mudah dibaca.

In [80]:
# If Statements as Guards
month = 5
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5 if month == 4:
    print("A weekday in April")
  case 1 | 2 | 3 | 4 | 5 if month == 5:
    print("A weekday in May")
  case _:
    print("No match")

A weekday in May


### Analisis:
Kode tersebut menunjukkan penggunaan guard conditions dalam struktur match–case, yaitu kondisi tambahan menggunakan if pada setiap case. Variabel day bernilai 4 dan month bernilai 5. Baris pertama mengecek apakah day termasuk 1–5 dan month == 4, namun syarat bulan tidak terpenuhi. Baris kedua juga memeriksa apakah day termasuk 1–5 dan month == 5, dan karena kedua kondisi benar, program mencetak "A weekday in May". Jika tidak ada case yang cocok, barulah case _ akan dieksekusi sebagai default.